modules/aaudio/recorder.c

```diff
@@ -0,0 +1,200 @@
+/**
+ * @file aaudio/recorder.c  AAudio audio driver for Android
+ *
+ * Copyright (C) 2024 Juha Heinanen
+ * Copyright (C) 2024 Sebastian Reimers
+ */
+
+#include <re.h>
+#include <rem.h>
+#include <baresip.h>
+#include <string.h>
+
+#include "aaudio.h"
+
+
+struct ausrc_st {
+	struct ausrc_prm src_prm;
+	ausrc_read_h *rh;
+	ausrc_error_h *errh;
+	void   *sampv;
+	size_t  sampsz;
+	size_t  sampc;
+	uint64_t samps;
+	void *arg;
+};
+
+
+AAudioStream *recorderStream = NULL;
+
+
+static void ausrc_destructor(void *arg)
+{
+	struct ausrc_st *st = arg;
+
+	close_stream(recorderStream);
+
+	mem_deref(st->sampv);
+	st->rh = NULL;
+	st->errh = NULL;
+}
+
+
+/**
+ * For an input stream, this function should read and process numFrames of
+ * data from the audioData buffer. The data in the audioData buffer must not
+ * be modified directly. Instead, it should be copied to another buffer
+ * before doing any modification. Note that numFrames can vary unless
+ * AAudioStreamBuilder_setFramesPerDataCallback() is called. Not currently
+ * called.
+ */
+ static int dataCallback(AAudioStream *stream, void *userData,
+			 void *audioData, int32_t numFrames) {
+	(void)stream;
+	struct ausrc_st *st = userData;
+	struct auframe af;
+
+	size_t sampc = 0;
+
+	sampc = numFrames;
+	if (sampc > st->sampc) {
+		st->sampv = mem_realloc(st->sampv, st->sampsz * sampc);
+		st->sampc = sampc;
+	}
+
+	if (!st->sampv)
+		return ENOMEM;
+
+	auframe_init(&af, st->src_prm.fmt, st->sampv, sampc,
+		     st->src_prm.srate, st->src_prm.ch);
+
+	memcpy(st->sampv, audioData, auframe_size(&af) * af.ch);
+
+	af.timestamp = st->samps * AUDIO_TIMEBASE /
+		       (st->src_prm.srate * st->src_prm.ch);
+	st->samps += sampc;
+	st->rh(&af, st->arg);
+
+	return 0;
+}
+
+
+static int open_recorder_stream(struct ausrc_st *st) {
+
+	AAudioStreamBuilder *builder;
+	aaudio_result_t result;
+
+	result = AAudio_createStreamBuilder(&builder);
+	if (result != AAUDIO_OK) {
+		warning("aaudio: recorder: failed to create stream builder: "
+			"error %s\n", AAudio_convertResultToText(result));
+		return result;
+	}
+
+	AAudioStreamBuilder_setDirection(builder, AAUDIO_DIRECTION_INPUT);
+	AAudioStreamBuilder_setSharingMode(builder,
+		AAUDIO_SHARING_MODE_EXCLUSIVE);
```

Comment:
Maybe better use `AAUDIO_SHARING_MODE_SHARED`?

Reply:
You mean allow the microphone to be used at the same time by many apps?  Anyway, I checked and the exclusive mode was not accepted and the result is shared.